### DISCUSSION TITLE
SVGGraphics2D support for TexturePaint (Generate Fill Patterns and Raster Bitmap images)

### DISCUSSION THREAD

**Discussion by istinnstudio** [jfree/jfreesvg#49] [Ideas]
This a feature request/idea, I am using successfully jfreesvg to generate SVGs from Java2D drawing API. Thanks for your efforts in this library. As far as I can see, TexturePaint objects with raster image references cannot be translated. I was wondering if it is relatively easy to implement some of the TexturePaint object elements like filling shapes with raster images. 

Example of SVGs with raster fills here: 
https://stackoverflow.com/questions/3796025/fill-svg-path-element-with-a-background-image
https://stackoverflow.com/questions/53878582/fill-a-svg-path-with-a-full-background-image

A TexturePaint object through SVGGraphics2D should generate SVGs with the image xlink:href elements involved, as appearing in java2D code, generating the reference link to the local raster file(s), or url(s) if any. Then someone could copy the raster files involved  and paste them near the generated SVG in order use them as a "group of files referenced from the SVG", edit etc.
I am not sure though (not tested), if other vector fill patterns painted through java2D TexturePaint are supported from SVGGraphics2D. 
http://tutorials.jenkov.com/svg/fill-patterns.html
Seems that there is no error message when using TexturePaint with SVGGraphics2D, but the image references are not being generated.

EXAMPLE SVG with A LOCAL FILE REFERENCE, could be any image
```
<?xml version="1.0" encoding="UTF-8"?>
<svg version="1.1" baseProfile="full" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:ev="http://www.w3.org/2001/xml-events">
    <title>Text Pattern Fill Example</title>
    <defs>
        <pattern id="img1" patternUnits="userSpaceOnUse" width="600" height="450">
            <image xlink:href="file:///C:/svg_test/0.png" x="0" y="0" width="600" height="450"/><!-- Image from ... -->
        </pattern>
    </defs>
    <path d="M5,50              l0,100 l100,0 l0,-100 l-100,0              M215,100              a50,50 0 1 1 -100,0 50,50 0 1 1 100,0              M265,50              l50,100 l-100,0 l50,-100              z" fill="url(#img1)"/>
        
</svg>
```

**Comment by jfree** [comment]
Thanks for the feature request and especially the extra links.  You are right that JFreeSVG is not supporting `TexturePaint` and since it is my aim to fully support the Java2D API as far as possible, I'll take a look at this.

**Reply by istinnstudio** [reply]
this is great, so if and whenever, this new feature arrives at any functional level, I'm willing to test and provide results  if necessary.

**Reply by istinnstudio** [reply]
I would like to note that Batik SVGGenerator supports texturepaint class embedding raster images as base64 binary, part of the SVG specification. It works perfectly on every browser and inkscape.  

a simple seamless background then is possible and it is fast.

java code to be translated to svg looks like this:

```
             myBackgroundImageFile = ImageIO.read(new File(Global.getBackgroundTileFilePath()));
            } catch (IOException e) {
               e.printStackTrace();
            }
            TexturePaint tp = new TexturePaint(myBackgroundImageFile, new Rectangle(0, 0, myBackgroundImageFile.getWidth(), myBackgroundImageFile.getHeight()));

svggraphics2d.setPaint(tp);
```

Of course this does not currently work in jfreesvg, considering jfreesvg can be much faster on complex shapes,  this would be a very useful addition.
So if you can take a look at Batik's approach maybe it is possible to port this feature in jfreesvg, texturepaints and in general raster images as "base64" tags.

example:

```
  ><defs id="defs1"
    ><pattern width="1" x="0" id="pattern1" height="1" y="0" patternUnits="userSpaceOnUse"
      ><use xlink:href="#image1"
      /></pattern
      ><filter width="100%" x="0%" height="100%" y="0%" id="alphaCompositeXor" filterUnits="objectBoundingBox"
      ><feFlood style="flood-color:white; flood-opacity:1;" result="flood"
        /><feComposite k2="0" in2="SourceGraphic" operator="xor" result="composite" in="BackgroundImage"
        /><feMerge
        ><feMergeNode in="flood"
          /><feMergeNode in="composite"
        /></feMerge
      ></filter
      ><image width="1" id="image1" height="1" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAAC0lEQVR4XmNgAAIA
AAUAAQYUdaMAAAAASUVORK5CYII="
    /></defs
```

**Reply by jfree** [reply]
Thanks for the follow-up.  I'm going to add a test case or two for this in https://github.com/jfree/graphics2d-tester then hopefully find enough time to add the feature.

**Comment by istinnstudio** [comment]
thanks for taking this into account
